test/test_server_trio.py

```diff
@@ -0,0 +1,189 @@
+import functools
+import logging
+
+import pytest
+import trio
+
+from pymodbus.client.asynchronous.schedulers import TRIO
+from pymodbus.client.asynchronous.tcp import AsyncModbusTCPClient
+from pymodbus.datastore.context import ModbusServerContext, ModbusSlaveContext
+from pymodbus.device import ModbusDeviceIdentification
+from pymodbus.pdu import ExceptionResponse, ModbusExceptions
+from pymodbus.register_read_message import ReadHoldingRegistersResponse
+from pymodbus.server.trio import tcp_server
+from pymodbus.register_write_message import WriteMultipleRegistersResponse
+
+
+class RunningTrioServer:
+    def __init__(self, context, host, port):
+        self.context = context
+        self.host = host
+        self.port = port
+
+
+@pytest.fixture(name="trio_server_context")
+async def trio_server_context_fixture(request):
+    node = request.node
+    single = node.get_closest_marker("single")
+    no_contexts = node.get_closest_marker("no_contexts")
+
+    if no_contexts:
+        slaves = {}
+    else:
+        slave_context = ModbusSlaveContext()
+
+        if single:
+            slaves = slave_context
+        else:
+            slaves = {0: slave_context}
+
+    server_context = ModbusServerContext(slaves=slaves, single=single)
+
+    return server_context
+
+
+@pytest.fixture(name="trio_tcp_server")
+async def trio_tcp_server_fixture(request, nursery, trio_server_context):
+    node = request.node
+    broadcast_enable = node.get_closest_marker("broadcast_enable")
+    ignore_missing_slaves = node.get_closest_marker("ignore_missing_slaves")
+
+    host = "127.0.0.1"
+
+    identity = ModbusDeviceIdentification()
+
+    [listener] = await nursery.start(
+        functools.partial(
+            trio.serve_tcp,
+            functools.partial(
+                tcp_server,
+                context=trio_server_context,
+                identity=identity,
+                ignore_missing_slaves=ignore_missing_slaves,
+                broadcast_enable=broadcast_enable,
+            ),
+            host=host,
+            port=0,
+        ),
+    )
+
+    yield RunningTrioServer(
+        context=trio_server_context,
+        host=host,
+        port=listener.socket.getsockname()[1],
+    )
+
+
+@pytest.fixture(name="trio_tcp_client")
+async def trio_tcp_client_fixture(trio_tcp_server):
+    modbus_client = AsyncModbusTCPClient(
+        scheduler=TRIO,
+        host=trio_tcp_server.host,
+        port=trio_tcp_server.port,
+    )
+
+    async with modbus_client.manage_connection() as protocol:
+        yield protocol
+
+
+@pytest.mark.trio
+async def test_read_holding_registers(trio_tcp_client, trio_tcp_server):
```

Comment:
These tests looks more like functional tests than unit tests. I can move this to functional tests later to #2 when functional tests actually are in place. Also the test seems to be failing on mac osx with ` trio.TrioInternalError: internal error in Trio - please file a bug!`.

Reply:
I added a bunch more unity tests as well.